### PR TITLE
fix(raft): notify role change listener only when transition completed

### LINKED ISSUE #7862: Appender position is smaller than previous appender position
**Describe the bug**

During the start-up procedure, the role change listener might be called twice. Basically, when a broker transitions to the leader role, its transition completes when the initial entry gets committed. When committing the initial entry, the partition becomes ready so that the Zeebe partition services (like Log Stream) are started which triggers: 

https://github.com/camunda-cloud/zeebe/blob/c844986457d3d6466a2b43e64f9a92161184ad5d/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartition.java#L297-L302

First, it will add a role change listener to the Raft partition, and second, it will execute Zeebe transition by calling `#onRoleChange()` which triggers starting the Zeebe partition services. Starting the Zeebe partition services may happen before the actual initial entry got committed eventually. As a consequence, Zeebe transitions twice during the start-up procedure:

1. First time by calling directly `ZeebePartition#onRoleChange()` and actively querying Raft's current role (thus bypassing the listener notification)
2. Second time when the initial entry gets committed eventually, which notifies the registered role change listener

As a consequence, with calling `#onRoleChange()` the first time, the write buffer (i.e., dispatcher, Log Stream) gets initialized with the last committed application entry. When the role change listener is notified, the write buffer is closed and again opened with an initial position which again is the last committed application entry. However, in between, there might already be commands/events being received and replicated but _not_ committed yet. This results in initializing the write buffer with a lower (and committed) initial position than the appender position in `LeaderAppender#lastZbEntry`.

**Expected behavior:**
When transitioning to the leader role, the role change listeners are called only once when the initial entry gets committed.

**Log/Stacktrace**
```
java.lang.IllegalStateException: Unexpected position 1985209002 was encountered after position 1985209021 when appending positions <1985209002, 1985209003>.
  at io.atomix.raft.roles.LeaderRole.safeAppendEntry(LeaderRole.java:567) ~[zeebe-atomix-cluster-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT]
  at io.atomix.raft.roles.LeaderRole.lambda$appendEntry$8(LeaderRole.java:547) ~[zeebe-atomix-cluster-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT]
  at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:174) [zeebe-atomix-utils-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT]
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) [?:?]
  at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
  at java.lang.Thread.run(Thread.java:833) [?:?]
```
---
*Initial issue description:*

**Describe the bug**

It seems the node [becomes leader](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-09-17T11:23:35.900610000Z&dateRangeStart=2021-09-17T10:24:05.766Z&dateRangeEnd=2021-09-17T12:24:05.766Z&project=camunda-cloud-240911&minLogLevel=0&customFacets=resource.labels.pod_name,jsonPayload.context.actor-name&limitCustomFacetWidth=true&advancedFilter=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22development-worker-1%22%0Aresource.labels.namespace_name%3D%2285c25a42-8853-48c3-be66-9c93d7c67eb7-zeebe%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-3%22&interval=JUMP_TO_TIME&scrollTimestamp=2021-09-17T11:23:31.471036000Z&pinnedLogId=u2rvtatnqfk3r7y2&pinnedLogTimestamp=2021-09-17T11:23:31.412949Z) for partition 4 and the [appender fails with: ` Unexpected position 147711 was encountered after position 147720 when appending positions <147711, 147716>.`](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-09-17T11:23:35.900610000Z&dateRangeStart=2021-09-17T10:24:05.766Z&dateRangeEnd=2021-09-17T12:24:05.766Z&project=camunda-cloud-240911&minLogLevel=0&customFacets=resource.labels.pod_name,jsonPayload.context.actor-name&limitCustomFacetWidth=true&advancedFilter=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22development-worker-1%22%0Aresource.labels.namespace_name%3D%2285c25a42-8853-48c3-be66-9c93d7c67eb7-zeebe%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-3%22&interval=JUMP_TO_TIME&scrollTimestamp=2021-09-17T11:23:35.737032000Z&pinnedLogId=4dmr7inl310nclvr&pinnedLogTimestamp=2021-09-17T11:23:35.737032Z)

[The appender is then closed](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-09-17T11:23:35.900610000Z&dateRangeStart=2021-09-17T10:24:05.766Z&dateRangeEnd=2021-09-17T12:24:05.766Z&project=camunda-cloud-240911&minLogLevel=0&customFacets=resource.labels.pod_name,jsonPayload.context.actor-name&limitCustomFacetWidth=true&advancedFilter=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22development-worker-1%22%0Aresource.labels.namespace_name%3D%2285c25a42-8853-48c3-be66-9c93d7c67eb7-zeebe%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-3%22&interval=JUMP_TO_TIME&scrollTimestamp=2021-09-17T11:23:35.748024000Z&pinnedLogId=r8kcepu43p88ec4m&pinnedLogTimestamp=2021-09-17T11:23:35.742511Z) which causes then later errors like [retrieving the write buffer ](https://console.cloud.google.com/logs/viewer?expandAll=false&timestamp=2021-09-17T11:23:35.900610000Z&dateRangeStart=2021-09-17T10:24:05.766Z&dateRangeEnd=2021-09-17T12:24:05.766Z&project=camunda-cloud-240911&minLogLevel=0&customFacets=resource.labels.pod_name,jsonPayload.context.actor-name&limitCustomFacetWidth=true&advancedFilter=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.cluster_name%3D%22development-worker-1%22%0Aresource.labels.namespace_name%3D%2285c25a42-8853-48c3-be66-9c93d7c67eb7-zeebe%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-3%22&interval=JUMP_TO_TIME&scrollTimestamp=2021-09-17T11:23:35.766296000Z&pinnedLogId=xlqyovpgvycecls9&pinnedLogTimestamp=2021-09-17T11:23:35.766296Z)

Related error groups:

 * https://console.cloud.google.com/errors/CJX23aTNsObURA?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911
 * https://console.cloud.google.com/errors/CPDYp5qk36X1GA?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911
 * https://console.cloud.google.com/errors/CNG7qLn5-ObZwwE?service=zeebe&time=P7D&refresh=off&project=camunda-cloud-240911

**To Reproduce**

**Expected behavior**

No problems on appending w.r.t the position.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
java.lang.RuntimeException: Actor is closed
at io.camunda.zeebe.logstreams.impl.log.LogStreamImpl.newLogStreamRecordWriter (LogStreamImpl.java:99)
at io.camunda.zeebe.broker.system.management.LeaderManagementRequestHandler.lambda$onBecomingLeader$2 (LeaderManagementRequestHandler.java:75)
at io.camunda.zeebe.util.sched.ActorJob.invoke (ActorJob.java:73)
at io.camunda.zeebe.util.sched.ActorJob.execute (ActorJob.java:39)
at io.camunda.zeebe.util.sched.ActorTask.execute (ActorTask.java:122)
at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask (ActorThread.java:94)
at io.camunda.zeebe.util.sched.ActorThread.doWork (ActorThread.java:78)
at io.camunda.zeebe.util.sched.ActorThread.run (ActorThread.java:191)
```

</p>
</details>

**Environment:**
 - OS: 
 - Zeebe Version: 1.2.0-alpha2
 - Configuration:

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

* Removes calling `ZeebePartition#onRoleChange()` when registering the role change listener
* Raft notifies the role change listeners only when the transition is completed, and in the case of a leader transition, the initial entry gets committed
* Adjust some test cases to ensure that the actual test scenario is tested

## Related issues

closes #7862

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by romansmirnov** [comment]
@deepthidevaki,

This PR implements the potential solution we have discussed to avoid calling the role change listener twice. While the solution works in general, I still wondering if there are any alternative solutions you could think of?

Why am I asking? With changing the behavior of `RaftContext#addRoleChangeListener()`, i.e., submitting a command that runs with the Raft executor asynchronously which then adds the listener and executes the listener directly if no transition is currently ongoing, but now the semantic of some test cases might have changed. For example, the following test case:

https://github.com/camunda-cloud/zeebe/blob/b0413b9c03bbb8bce3b4e8f9c923e1b5343e9c85/atomix/cluster/src/test/java/io/atomix/raft/RaftTest.java#L344-L360

Before that change, it behaved pretty deterministically:
1. Create 3 Raft Servers so that a leader gets elected
2. Get the Leader
3. Add Role Change Listener (they are directly added to the set of listeners)
4. Let the leader step down
5. Await that a new leader gets elected and once the initial entry gets committed the listener is called

With the PR, now the behavior changes in a way that the listener gets called by the previous leader and not by the new leader. So while the test case itself is still fine - it still gets green, the code change potentially changes the semantic of this (and maybe other) test case(s).

Of course, I could now go through the test cases and adjust them accordingly 😄 But maybe there is any alternative you could think of?

**Comment by romansmirnov** [comment]
@deepthidevaki,

What do you think about the following approach? Basically, instead of pushing the case that the listener might be called twice the Raft layer, the concern/issue should stay within the broker itself. Meaning, keep the way of registering and triggering the listener as it is in `ZeebePartition`, but handle the case more gracefully that a transition to the LEADER role is triggered twice here:

https://github.com/camunda-cloud/zeebe/blob/b0413b9c03bbb8bce3b4e8f9c923e1b5343e9c85/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/PartitionTransitionImpl.java#L64-L91

Basically, if the current role is equal to the role to transition to, then do nothing and return a completed future. In my opinion, this should not cause any issues also when calling follow-up listener like `PartitionContext#notifyListenersOf...()`.

That way, we keep the "issue" in the broker and don't push it to the Raft layer.

What do you think?

**Comment by deepthidevaki** [comment]
Having a check in ZeebePartition to prevent multiple transitions to the same Role is a good idea. But I think this will not solve the issue completely. This will make sure that the second transition will not happen. But the main problem here is that the ZeebePartition is transitioning to Leader before InitialEntry is committed, because it actively queries Raft's current role, thus bypassing the listener notification. So in this case it can still happen that in the first transition, it reads the wrong lastEntry, because the entries are not committed yet. This will leads to the same error before the second transition happens. So we have to anyway remove the following call 
```
 onRoleChange(context.getRaftPartition().getRole(), context.getRaftPartition().term());
```
from `ZeebePartition#registerListenersAndTriggerRoleChange`

**Comment by romansmirnov** [comment]
@deepthidevaki, true, you are right!

**Comment by deepthidevaki** [comment]
> Of course, I could now go through the test cases and adjust them accordingly smile But maybe there is an alternative you could think of?

I cannot think of alternative solutions. Are there too many tests that needs to be adjusted because of this change? 
It is ok to adjust test, but we should ensure that we are not breaking any critical guarantees.

**Comment by romansmirnov** [comment]
@deepthidevaki, thanks for your input for the quick discussion. Just a quick summary of what, we have talked about:

* For now, we will not consider the following potential solution: Adding an additional guard in `PartitionTransitionImpl#transitionTo()` to abort the transition request when the current role is the same as the requested role. While in general, this guard could make sense it would require additional investigation to understand its impact to ensure that all Zeebe partition services are started correctly and only half of them.
* We will continue considering the solution when adding a role change listener to call the listener immediately if there is no ongoing transition. That way, the `ZeebePartition` does not need to trigger a role change on its own, and they will be notified once the transition is finished and in the case of the leader, the initial entry gets committed.

**Comment by romansmirnov** [comment]
@deepthidevaki,

I am done with the implementation of the fix. I tried to write a test case, therefore, but I couldn't find a way to reproduce the scenario in a test case and to control for example when the commit happens should happen. I think, the current test cases (i.e., the one that I adjusted a bit) implicitly ensure that the listener is notified correctly and especially in the case of a leader when the initial entry gets committed.

However, if you have an approach, I am happy to receive it!

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
:rocket: LGTM. 
May be we can also add a test to verify that the listener is notified after it is added, even if there was no role change. This is the semantics changed in this PR, right.

**Comment by romansmirnov** [comment]
@deepthidevaki, sure thing!

**Review by deepthidevaki — APPROVED** [review]
:rocket:  Thanks.

**Comment by romansmirnov** [comment]
bors merge
